### PR TITLE
【PaddlePaddle Hackathon】1、在 Paddle 中新增 AdaptiveLogSoftmaxWithLoss

### PR DESCRIPTION
**Pull request by gsq7474741**
### PR types

New features
### PR changes

APIs
### Describe

- Add nn.AdaptiveLogSoftmaxWithLoss, with unittest.
- Task #36009

### CODE REVIEW

**Comment by gsq7474741** [comment]
@TCChenlong

**Comment by gsq7474741** [comment]
@zhiboniu 您好，请问iscan的报错怎么看呀，点进ditails看不到

File: python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import paddle
+from paddle import nn
+from paddle.nn import functional as F
+
+np.random.seed(10)
+paddle.seed(10)
+
+
+class TestNNAdaptiveLogSoftmaxWithLossAPI(unittest.TestCase):
+    def test_adaptive_log_softmax(self):
```
Comment (zhiboniu) [inline python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py]: 不同的测试可以分开不同的test函数，更容易区分。

File: python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+        self.assertEqual(asfm(x, y).output.shape, [128])
+
+        # log_probs actually returns log_proba
+        asfm = nn.AdaptiveLogSoftmaxWithLoss(8, 4, [2], div_value=2.)
+        x = paddle.randn((4, 8))
+        logprob_out = asfm.log_prob(x)
+        np.testing.assert_array_almost_equal(
+            paddle.exp(logprob_out).sum(1), paddle.ones([4]))
+        # if_equal=(paddle.abs(paddle.exp(logprob_out).sum(1)-paddle.ones([4]))<paddle.ones([4])*1e-5).numpy().tolist()
```
Comment (zhiboniu) [inline python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py]: 注释掉的没用的代码需要清理掉，除非是说明或者有其他目的

File: python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        asfm.head.weight.detach().abs()
+        asfm.head.bias.detach().abs()
+        asfm.head.weight.detach()[asfm.shortlist_size:, :] *= 0.
+
+        out = asfm.predict(x)
+        np.testing.assert_array_almost_equal(
+            out, asfm.log_prob(x).argmax(axis=1))
+
+        # argmax outside of shortlist
+        asfm = nn.AdaptiveLogSoftmaxWithLoss(
```
Comment (zhiboniu) [inline python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py:113]: 需要添加几个测试案例验证计算结果的正确性

File: python/paddle/nn/layer/loss.py [hunk 4]
```diff
@@ -15,14 +15,276 @@
...
+from .. import Layer, Sequential, LayerList
+from paddle.nn.layer import Linear
+from paddle import Tensor
 
 __all__ = []
+_ASMoutput = namedtuple('_ASMoutput', ['output', 'loss'])
+
+
+class AdaptiveLogSoftmaxWithLoss(Layer):
+    r"""Efficient softmax approximation as described in
```
Comment (zhiboniu) [inline python/paddle/nn/layer/loss.py]: 应尽量避免直接copy其他开源的代码。
如无法避免，也应检查源码的开源协议是否合适并在函数头添加引用源及原始协议注释
这里源码应该是BSD协议吧

File: python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py (hunk 3 above)
Comment (gsq7474741) [reply python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py:113]: @zhiboniu 我看PaddleTest仓库里那些测试都是写死的验证点，这个我试了一下，发现数值不稳定，就没有添加写死的test case

**Comment by gsq7474741** [comment]
@zhiboniu 已更新

File: python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            8, 10, [4, 8], div_value=2., head_bias=True)
```
Comment (zhiboniu) [inline python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py:114]: 你这样还是在用自己的代码验证自己代码的正确性，是不够的。
你可以参考linear等网络层的测试方式，固定网络层的随机参数数值，就可以验证输出数值的正确性。也可以添加跟普通softmax数值的对比结果。

Comment (gsq7474741) [reply python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py:114]: ok下午试一下

Comment (gsq7474741) [reply python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py:114]: @zhiboniu 已更新

File: python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+import numpy as np
+import paddle
+from paddle import nn
+from paddle.nn import functional as F
+
+
+class TestNNAdaptiveLogSoftmaxWithLossAPI(unittest.TestCase):
+    def setUp(self):
+        paddle.disable_static()
+        paddle.seed(10)
```
Comment (zhiboniu) [inline python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py]: 把seed去掉random就不会挂了

Comment (gsq7474741) [reply python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py]: 好滴 我试一下

Comment (gsq7474741) [reply python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py]: @zhiboniu 已更新

File: python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        asfm.head.weight.detach()[asfm.shortlist_size:, :] *= 0.
+
+        out = asfm.predict(x)
+        np.testing.assert_array_almost_equal(
+            out, asfm.log_prob(x).argmax(axis=1))
+
+        # argmax outside of shortlist
+        asfm = nn.AdaptiveLogSoftmaxWithLoss(
+            8, 10, [4, 8], div_value=2., head_bias=True)
+        asfm.head.weight.detach().abs()
```
Comment (zhiboniu) [inline python/paddle/fluid/tests/unittests/test_adaptive_log_softmax_with_loss.py:115]: 参考上面两个对话，你当前还是没有能验证实现的正确性。
以test_linear.py为例，可以按如下步骤完成：
1）固定可学习参数权重值为1。
2）获得模型推理结果概率分布，不仅仅是argmax值。
3）使用numpy实现推理公式。
4）使用同一个随机输入验证paddle输出概率分布与numpy计算结果相同。
